frame/contracts/src/tests.rs

```diff
@@ -2966,6 +2983,83 @@ fn sr25519_verify() {
 	})
 }
 
+#[test]
+fn test_failed_charge_should_roll_back_call() {
+	let (wasm_caller, _) = compile_module::<Test>("call_runtime_and_call").unwrap();
+	let (wasm_callee, _) = compile_module::<Test>("store").unwrap();
+
+	ExtBuilder::default().existential_deposit(200).build().execute_with(|| {
+		let _ = Balances::deposit_creating(&ALICE, 1_000_000);
+
+		// Instantiate both contracts.
+		let addr_caller = Contracts::bare_instantiate(
+			ALICE,
+			0,
+			GAS_LIMIT,
+			None,
+			Code::Upload(wasm_caller),
+			vec![],
+			vec![],
+			false,
+		)
+		.result
+		.unwrap()
+		.account_id;
+		let addr_callee = Contracts::bare_instantiate(
+			ALICE,
+			0,
+			GAS_LIMIT,
+			None,
+			Code::Upload(wasm_callee),
+			vec![],
+			vec![],
+			false,
+		)
+		.result
+		.unwrap()
+		.account_id;
+
+		// Give caller proxy access to Alice.
+		assert_ok!(Proxy::add_proxy(RuntimeOrigin::signed(ALICE), addr_caller.clone(), (), 0));
+
+		// Create a Proxy call that will attempt to transfer away's Alice balance, so we can test a
+		// failed charge.
+		let min_balance = <Test as Config>::Currency::minimum_balance();
+		let transfer_call =
+			Box::new(RuntimeCall::Balances(pallet_balances::Call::transfer_allow_death {
+				dest: CHARLIE,
+				value: pallet_balances::Pallet::<Test>::free_balance(&ALICE) - min_balance,
+			}));
+
+		// Wrap the transfer call in a proxy call.
+		let transfer_proxy_call = RuntimeCall::Proxy(pallet_proxy::Call::proxy {
+			real: ALICE,
+			force_proxy_type: Some(()),
+			call: transfer_call,
+		});
+
+		// bump the deposit per byte to a high value to trigger a FundsUnavailable error.
+		DEPOSIT_PER_BYTE.with(|c| *c.borrow_mut() = 100);
+
+		let data = (
+			100u32, // storage length
+			addr_callee,
+			transfer_proxy_call,
+		);
+
+		let result = <Pallet<Test>>::call(
+			RuntimeOrigin::signed(ALICE),
+			addr_caller.clone(),
+			0,
+			GAS_LIMIT,
+			None,
+			data.encode(),
+		);
+
+		assert_err_ignore_postinfo!(result, TokenError::FundsUnavailable);
```

Comment:
would be nice to test that the transaction gas were actually charge to the user, but I am not too sure how this can be done within our test framework

Reply:
You can check the events emitted during execution. We do that in other tests. There is a event for transaction fee payment.

Reply:
we would need the transaction_payment pallet for this in test, I don't think we have it 🤔  might have overlooked, will check again

Reply:
Ahh yeah. You are right. We didn't set this up in tests. It is fine. We don't really need to test this.